src/pids.c

```diff
@@ -96,6 +97,99 @@ char *utf8_encode(int encoding, char *buf, int len)
     return NULL;
 }
 
+void report(pids_t *st)
+{
+    int i;
+    char *country_code = NULL;
+    char *name = NULL;
+    char *slogan = NULL;
+    char *message = NULL;
+    char *alert = NULL;
+    float latitude = NAN;
+    float longitude = NAN;
+    int altitude = 0;
+    nrsc5_sis_asd_t *audio_services = NULL;
+    nrsc5_sis_dsd_t *data_services = NULL;
+
+    if (st->country_code[0] != 0)
+        country_code = st->country_code;
+
+    if (st->short_name[0] != 0)
+        name = st->short_name;
+
+    if (st->slogan_displayed)
+        slogan = utf8_encode(st->slogan_encoding, st->slogan, st->slogan_len);
+    else if (st->long_name_displayed)
+        slogan = st->long_name;
+
+    if (st->message_displayed)
+        message = utf8_encode(st->message_encoding, st->message, st->message_len);
+
+    if (st->alert_displayed)
+    {
+        int cnt_bytes = 1 + (2 * st->alert_cnt_len);
+        alert = utf8_encode(st->alert_encoding, st->alert + cnt_bytes, st->alert_len - cnt_bytes);
+    }
+
+    if (!isnan(st->latitude) && !isnan(st->longitude))
+    {
+        latitude = st->latitude;
+        longitude = st->longitude;
+        altitude = st->altitude;
+    }
+
+    for (i = MAX_AUDIO_SERVICES - 1; i >= 0; i--)
+    {
+        if (st->audio_services[i].type != -1)
+        {
+            nrsc5_sis_asd_t *asd = malloc(sizeof(nrsc5_sis_asd_t));
+            asd->next = audio_services;
+            asd->program = i;
+            asd->access = st->audio_services[i].access;
+            asd->type = st->audio_services[i].type;
+            asd->sound_exp = st->audio_services[i].sound_exp;
+            audio_services = asd;
+        }
+    }
+
+    for (i = MAX_DATA_SERVICES - 1; i >= 0; i--)
+    {
+        if (st->data_services[i].type != -1)
+        {
+            nrsc5_sis_dsd_t *dsd = malloc(sizeof(nrsc5_sis_dsd_t));
+            dsd->next = data_services;
+            dsd->access = st->data_services[i].access;
+            dsd->type = st->data_services[i].type;
+            dsd->mime_type = st->data_services[i].mime_type;
+            data_services = dsd;
+        }
+    }
+
+    nrsc5_report_sis(st->input->radio, country_code, st->fcc_facility_id, name, slogan, message, alert,
+                     latitude, longitude, altitude, audio_services, data_services);
+
+    if (st->slogan_displayed)
```

Comment:
I think it would be generally better if you could just always call ```free``` on these pointers. I understand why you can't because of slogan is sometimes a pointer that needs freed and sometimes it is a pointer that should not be freed.

Maybe you should have another variable (which is ```const char *``` and never freed) that contains either long name or slogan. That way you can just always free these pointers. I'm not committed to this, just don't like the idea of a pointer sometimes being freeable and sometimes not freeable.

Reply:
I added a `strdup` to the `long_message` case, and made the `free`s below unconditional.